lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs

```diff
@@ -505,25 +511,49 @@ monitorStakePools tr gp nl db@DBLayer{..} = do
         -> NonEmpty (CardanoBlock sc)
         -> (BlockHeader, ProtocolParameters)
         -> IO (FollowAction ())
-    forward latestGarbageCollectionEpochRef blocks (_nodeTip, _pparams) = do
-        atomically $ forM_ blocks $ \case
+    forward latestGarbageCollectionEpochRef blocks _ = do
+        atomically $ forML_ blocks forAllBlocks forLastBlock
+        pure Continue
+      where
+        forAllBlocks = \case
             BlockByron _ -> pure ()
             BlockShelley blk -> do
                 let (slot, certificates) = poolCertsFromShelleyBlock blk
                 let header = toShelleyBlockHeader getGenesisBlockHash blk
-                runExceptT (putPoolProduction header (getProducer blk))
-                    >>= \case
-                        Left e ->
-                            liftIO $ traceWith tr $ MsgErrProduction e
-                        Right () ->
-                            pure ()
+                handleErr (putPoolProduction header (getProducer blk))
                 unless (null certificates) $ do
                     -- Before adding new pool certificates to the database, we
                     -- first attempt to garbage collect pools that have already
                     -- retired.
                     garbageCollectPools slot latestGarbageCollectionEpochRef
                     putPoolCertificates slot certificates
-        pure Continue
+
+        forLastBlock = \case
+            BlockByron blk ->
+                putHeader (toByronBlockHeader gp blk)
+            BlockShelley blk ->
+                putHeader (toShelleyBlockHeader getGenesisBlockHash blk)
+
+        handleErr action = runExceptT action
+            >>= \case
+                Left e ->
+                    liftIO $ traceWith tr $ MsgErrProduction e
+                Right () ->
+                    pure ()
+
+        -- | Like 'forM_', except runs the second action for the last element as
+        -- well (in addition to the first action).
+        forML_ :: (Monad m)
+            => NonEmpty a
+            -> (a -> m b) -- ^ action to run for all elements
+            -> (a -> m c) -- ^ action to run for the last element
+            -> m ()
+        {-# INLINE forML_ #-}
+        forML_ ne a1 a2 = go (NE.toList ne)
```

Comment:
**Suggestion:** replace this abstraction with something simpler.

**Justification:**

Very short and cryptic names are generally reserved for functions that are very general in nature, such as `forM`, or `scanM` and so on. Since we tend to reuse these general functions a great deal, it doesn't matter that they have short names: the names become part of the lexicon that everyone gets used to over time. And because we have to read them so frequently, being short has an advantage.

But this function is specialized to a single use case. Having a cryptic name means that the reader has to go to the definition of the function to figure out what it's doing.

The overall design also seems a bit brittle: if we need to change the way we record block headers (for example), then it's likely we'll need to replace `forML_` with something else entirely.

**Proposal:**

Why not write something like this, instead:
```hs
forward latestGarbageCollectionEpochRef blocks _ = do
    atomically $ do
        recordPoolCertificates blocks
        recordBlockHeaders blocks
    pure Continue

```
And then just let `recordPoolCertificates` and `recordBlockHeaders` decide what they want to do with the blocks they are given?

So we could have something like this:
```hs
recordPoolCertificates blocks = do
    forM_ blocks $ \case
        BlockByron _ ->
            -- Do nothing, as pools did not exist in the Byron era.
            pure ()
        BlockShelley blk ->
            -- Do stuff.

recordBlockHeaders blocks =
    case NE.last blocks of
        BlockByron blk ->
            putHeader (toByronBlockHeader gp blk)
        BlockShelley blk ->
            putHeader (toShelleyBlockHeader getGenesisBlockHash blk)
```
Then there is no need to invent an extra combinator.

What do you think?

Reply:
As I already explained... I think it's best to explicitly avoid traversing the list twice. This is a hot loop.

> But this function is specialized to a single use case. Having a cryptic name means that the reader has to go to the definition of the function to figure out what it's doing.

Yes it is. I checked if there's something similar already, but I couldn't find it. If you want to suggest a different name, please do so.

> The overall design also seems a bit brittle: if we need to change the way we record block headers (for example), then it's likely we'll need to replace forML_ with something else entirely.

I don't think this is a problem. If the code changes in the future, this will be a simple change.

Reply:
Not a huge fan of the name `forML_` which is a bit cryptic, but I do agree that traversing the list one once is essential here (that's about 4M elements on mainnet!) 

Perhaps resorting to a slightly more transparent name could help make this clearer. A suggestion: `forAllAndLastM`?